File: code/algo/getoffer/go/3/weight_random.go
package main

import (
	"math/rand"
	"sort"
)

type WeightRandom struct {
	pre []int
}

func NewWeightRandom(w []int) WeightRandom {
	for i := 1; i < len(w); i++ {
		w[i] += w[i-1]
	}
	return WeightRandom{w}
}

func (s *WeightRandom) PickIndex() int {
	x := rand.Intn(s.pre[len(s.pre)-1]) + 1
	return sort.SearchInts(s.pre, x)
}

File: code/algo/getoffer/c++/3/weight_random.cpp
#include <algorithm>
#include <cstdlib>
#include <vector>

class Solution {
public:
  Solution(std::vector<int> &w) {
    for (int i = 1; i < w.size(); ++i) {
      w[i] += w[i - 1];
    }
    pre = w;
  }

  int PickIndex() {
    int x = rand() % pre.back() + 1;
    return std::lower_bound(pre.begin(), pre.end(), x) - pre.begin();
  }

private:
  std::vector<int> pre;
};
